Google.Api.Gax.Grpc/HeaderParameterExtractor.cs

```diff
@@ -0,0 +1,86 @@
+﻿/*
+ * Copyright 2021 Google Inc. All Rights Reserved.
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Google.Api.Gax.Grpc
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    public sealed class HeaderParameterExtractor<TRequest>
+    {
+        private readonly IReadOnlyList<ParamExtraction> _extractions;
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="paramName"></param>
+        /// <param name="regex"></param>
+        /// <param name="selector"></param>
+        public HeaderParameterExtractor(string paramName, Regex regex, Func<TRequest, string> selector)
+        {
+            _extractions = new List<ParamExtraction> { new ParamExtraction(paramName, regex, selector) };
+        }
+
+        private HeaderParameterExtractor(IEnumerable<ParamExtraction> extractions)
+        {
+            _extractions = new List<ParamExtraction>(extractions);
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="paramName"></param>
+        /// <param name="regex"></param>
+        /// <param name="selector"></param>
+        /// <returns></returns>
+        public HeaderParameterExtractor<TRequest> Add(string paramName, Regex regex, Func<TRequest, string> selector)
+        {
+            return new HeaderParameterExtractor<TRequest>(_extractions.Concat(new[] { new ParamExtraction(paramName, regex, selector) }));
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="request"></param>
+        /// <returns></returns>
+        public Dictionary<string, string> Extract(TRequest request)
+        {
+            return _extractions.GroupBy(pe => pe.ParamName)
```

Comment:
This is the bit that feels like it can be done once on construction instead of repeatedly. It may require keeping more state than we would otherwise do (so that we've got the original extractions as well as the "processed" version) but I doubt that'll be a problem.

Reply:
👍